src/azul/terra.py

```diff
@@ -272,8 +283,23 @@ def run_sql(self, query: str) -> BigQueryRows:
                     sleep(delay)
                 else:
                     raise e
+            else:
+                if log.isEnabledFor(logging.DEBUG):
+                    log.debug('Job info: %r', self._job_info(result, job))
+                return result
         assert False
 
+    def _job_info(self, query_result: RowIterator, job: QueryJob) -> MutableJSON:
+        info = query_result.api_request(method='GET',
+                                        path=f'/projects/{self.credentials.project_id}/jobs/{job.job_id}')
+        info: MutableJSON = info['statistics']['query']
+        for key in ('referencedTables',
+                    'statementType',
+                    *(('queryPlan',) if config.debug < 2 else ())):
+            info.pop(key, None)
```

Comment:
Is it safe to modify this? If so, why?

Reply:
The response comes the `requests` library if you follow the call stack. So the JSON is coming from a server, and shouldn't connected to any other local objects. I can't find any official documentation that explicitly says it's safe to modify (the only docs I can find for `api_request` only go as far as to helpfully explain that it's the "The function to use to make API requests"). So I'm confident it works in practice, and can't imagine any reasonable circumstances in which it would stop working, but I can't rigorously guarantee that.

Reply:
I see. I didn't realize that this method makes another round-trip to BQ. It should be safe to modify the response from that. It's commonplace to assume that the return value is owned by the caller of a method, if that method returns a REST service response. However, it is always important to check for caching and sharing if the documentation is silent about ownership and mutability.

Having said that, the additional round-trip is surprising. Why is it necessary? Don't we already have the job info in memory somewhere?

Reply:
Sort of.

The `Job` object returned by BigQuery has _most_ of the info returned by this REST call, but it's not possible to fully recover the contents of https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#jobstatistics2 as requested in the ticket. For example, there is no reference to the `reservationUsage` field anywhere in the python API (I looked through the `dir` of the `Job` object and did a text search through `.venv/lib{64,}/python3.8/site-packages/google`). The REST call also has the advantage of already being is JSON format, sparing us the trouble of manually assembling it and extracting the structure of the nested `timeline` and `queryPlan` attributes. But the main motivation for making the second round trip is be thorough in collecting all of the fields.

Reply:
@hannes-ucsc  Pursuant to our parking lot discussion, I've found the point at which the Job object is constructed from the REST API response. Unfortunately, the JSON schema of this response is https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/query#response-body, which still lacks most of the info from https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#jobstatistics2.

Reply:
(The relevant library code is lines 1281-1289, `.venv/lib/python3.8/site-packages/google/cloud/bigquery/client.py`, in BigQuery v1.24.0)

Reply:
I don't think you're looking in the right places. The raw `/bigquery/v2/projects/{projectId}/jobs/{jobId}` response is in memory at this point:

![image](https://user-images.githubusercontent.com/5143256/102930909-3610a680-4452-11eb-9d67-634ff1af0c26.png)

and later, the bits we are interested in are still available:

![image](https://user-images.githubusercontent.com/5143256/102935579-4bd69980-445b-11eb-9ff2-efb0d23c4272.png)

Reply:
Instead of copy and pop, try a dict comprehension. And the conditional non-empty/empty tuple can just be an `or` in the `if` of that comprehension.